### PR TITLE
Create FlutterFragmentActivity

### PR DESCRIPTION
**Pull request by tvolkert**
This creates a `FlutterFragmentActivity` class that extends
the Android v4 Support librray's `FragmentActivity` class.
However, we intentionally do not bundle the support library
with our engine, so apps that wish to use this class are
responsible for including the support library .jar file in
their runtime deps when creating the final app.

flutter/flutter#10072

### CODE REVIEW

**Comment by jason-simmons** [comment]
LGTM

The license file needs to be updated to include the v4 support library classes

**Comment by collinjackson** [comment]
LGTM

It's sad that we are going to have to maintain a large block of boilerplate plugin registry forwarding code in both FlutterActivity and FlutterFragmentActivity (including things we haven't implemented yet, e.g. `onSaveInstanceState`). It's not very DRY and the potential for them to get out of sync is high. But I can't think of a better way right now.

This approach is fairly easy to understand from an end developer perspective at least. The only concern I have is that Android plugin authors might try to store registrar.activity() as a member with type FlutterActivity that they plan to use within the plugin when they need an Activity, Context, or FlutterView.Provider. Such a plugin would be unusable by FlutterFragmentActivity apps because registrar.activity() would returns a different object that doesn't extend FlutterActivity. I can't think of a solution to this problem right now though so I think we just have to put the onus on plugin authors not to assume that registrar.activity() returns a FlutterActivity.

**Comment by collinjackson** [comment]
Consider maybe adding a comment [here](https://github.com/flutter/engine/blob/master/shell/platform/android/io/flutter/plugin/common/PluginRegistry.java#L60) in PluginRegistry.java reminding plugin authors not to cast `registrar.activity()` a FlutterActivity or FlutterFragmentActivity unless they don't want to support the other.

File: shell/platform/android/io/flutter/app/FlutterFragmentActivity.java [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+import io.flutter.app.FlutterActivityDelegate.ViewFactory;
+import io.flutter.plugin.common.PluginRegistry;
+import io.flutter.plugin.common.PluginRegistry.Registrar;
+import io.flutter.view.FlutterView;
+
+/**
+ * Base class for activities that use Flutter who also require the use of the
+ * Android v4 Support library's {@link FragmentActivity}. Applications that
+ * don't have this need will likely want to use {@link FlutterActivity} instead.
+ * <p/>
```
Comment (Hixie) [inline shell/platform/android/io/flutter/app/FlutterFragmentActivity.java]: why the random empty paragraph? is that a javadoc thing?

**Comment by Hixie** [comment]
woah, we can't check in a .jar file.

File: tools/licenses/lib/main.dart [hunk 2]
```diff
@@ -2079,6 +2079,8 @@ class RepositoryFlutterDirectory extends RepositoryDirectory {
   RepositoryDirectory createSubdirectory(fs.Directory entry) {
     if (entry.name == 'sky')
       return new RepositoryExcludeSubpathDirectory(this, entry, const <String>['packages', 'sky_engine', 'LICENSE']); // that's the output of this script!
+    if (entry.name == 'shell')
+      return new RepositoryExcludeSubpathDirectory(this, entry, const <String>['platform', 'android', 'lib', 'android-support-v4.jar']); // not bundled or statically linked.
```
Comment (Hixie) [inline tools/licenses/lib/main.dart]: s/not/neither/, i assume

Comment (Hixie) [inline tools/licenses/lib/main.dart]: (and s/or/nor/)

**Comment by Hixie** [comment]
You'll want to pull this file down as part of the `gclient sync`, but other than that it looks fine to me. @abarth may notice things that I missed.

File: shell/platform/android/io/flutter/app/FlutterFragmentActivity.java (hunk 1 above)
Comment (tvolkert) [inline shell/platform/android/io/flutter/app/FlutterFragmentActivity.java]: Yeah, Javadoc isn't as smart about formatting as Dartdoc - with the exception of `{@foo}` directives, it treats everything as raw HTML, as in so doing, it collapses your paragraphs into one unless you add `<br/>` or `<p/>` tags.

**Comment by jason-simmons** [comment]
android-support-v4.jar should already be available in third_party/android_tools/sdk

**Pull request merged**